### ISSUE TITLE
base class inheritance for validations

### ISSUE THREAD

**Issue by roupen** [hanami/validations#114]
dry-validation has the ability to inherit a base schema as documented here:

http://dry-rb.org/gems/dry-validation/basics/working-with-schemas/

To allow for validations from a base schema to apply to the current and then in the block you can describe any additional validations. It doesn't seem like we have the ability to take advantage of this feature. It would probably involve doing something like:

```
class SpecialUser
  include Hanami::Validations::Form
  validations(BaseUser) do
    required(:extra_field)
  end
end
```

So basically being able to pass a schema (or ideally an entire other form much like you have support for composition in nested attributes)

**Comment by beauby** [comment]
An other option (the one I was outlining on the chat), would be to have the `validations` method behave like the `params` method in `Hanami::Controller`, in the sense that it would accept either a block for defining in-line schemas, or a schema object.

**Comment by solnic** [comment]
How about just having this:

``` ruby
class SpecialUser < BaseUser
  validations do
    required(:extra_field)
  end
end
```

?

**Comment by roupen** [comment]
@solnic Yes I think that would be the best solution!

**Comment by beauby** [comment]
@solnic That does look good indeed. For the record, is it currently working?

**Comment by solnic** [comment]
I don't know, try :)

**Comment by roupen** [comment]
I had already tried it didn't work :( when you do result.output it doesn't show all the attributes from the parent class which is why I was digging around to see how to accomplish inheritance. I believe the superclass' schema created by its validates block is overridden by the subclass's schema generated from its' validates block

**Comment by jodosha** [comment]
@roupen I tried to implement this feature, but it turned out that is an error prone solution. Because if we explicit allow inheritance, there are other things we should take care of for children classes: shared and inline predicates, error messages, and rules.

At this time, I'm afraid we can't manage it, but I keep track of it for the future.

**Comment by solnic** [comment]
Let's maybe pick this up once we're done with the model part. Inheritance is tricky in general, inheritance in validation objects is *even more tricky*. I'd like to see how we could improve dry-v itself to make inheritance easy-but-without-too-many-caveats - what we have already is pretty flexible, but I wouldn't consider it as "done".

**Comment by jodosha** [comment]
@roupen @solnic Agreed 👍
